crates/ide/src/folding_ranges.rs

```diff
@@ -146,8 +163,93 @@ pub(crate) fn folding_ranges(file: &SourceFile) -> Vec<Fold> {
     res
 }
 
-fn fold_kind(kind: SyntaxKind) -> Option<FoldKind> {
-    match kind {
+/// Builds a fold for the given syntax element.
+///
+/// This function creates a `Fold` object that represents a collapsible region in the code.
+/// If `collapsed_text` is enabled, it generates a preview text for certain fold kinds that
+/// shows a summarized version of the folded content.
+fn build_fold(element: &SyntaxElement, kind: FoldKind, collapsed_text: bool) -> Fold {
+    if !collapsed_text {
+        return Fold::new(element.text_range(), kind);
+    }
+
+    let fold_with_collapsed_text = match kind {
+        FoldKind::TailExpr => {
+            let expr = ast::Expr::cast(element.as_node().unwrap().clone()).unwrap();
+
+            let indent_level = expr.indent_level().0;
+            let indents = "    ".repeat(indent_level as usize);
+
+            let mut fold = Fold::new(element.text_range(), kind);
+            if let Some(collapsed_expr) = collapsed_text_from_expr(expr) {
+                fold = fold.with_text(format!("{indents}{collapsed_expr}"));
+            }
+            Some(fold)
+        }
+        FoldKind::Stmt => 'blk: {
+            let node = element.as_node().unwrap();
+
+            match_ast! {
+                match node {
+                    ast::ExprStmt(expr) => {
+                        let Some(expr) = expr.expr() else {
+                            break 'blk None;
+                        };
+
+                        let indent_level = expr.indent_level().0;
+                        let indents = "    ".repeat(indent_level as usize);
```

Comment:
Seemds odd to me that we need to add the leading whitespace, given the fold range starts at the statement start, not at the line start 🤔